File: src/main/java/io/github/seggan/jyxal/runtime/Lambda.java
package io.github.seggan.jyxal.runtime;

import java.lang.invoke.MethodHandle;
import java.util.ArrayList;
import java.util.List;

public record Lambda(int arity, MethodHandle handle) {

    public Object call(ProgramStack stack) {
        List<Object> args = new ArrayList<>();
        for (int i = 0; i < arity; i++) {
            args.add(stack.pop());
        }

        try {
            return handle.invoke(new ProgramStack(args));
        } catch (Throwable e) {
            throw new RuntimeException(e);
        }
    }

    public Object call(Object arg) {
        if (arity != 1) {
            throw new RuntimeException("Invalid arity");
        }
        try {
            return handle.invoke(new ProgramStack(arg));
        } catch (Throwable e) {
            throw new RuntimeException(e);
        }
    }
}

File: src/main/kotlin/io/github/seggan/jyxal/runtime/Lambda.kt
package io.github.seggan.jyxal.runtime

import java.lang.invoke.MethodHandle

data class Lambda(val arity: Int, val handle: MethodHandle) {

    fun call(stack: ProgramStack): Any {
        val args: MutableList<Any> = ArrayList()
        for (i in 0 until arity) {
            args.add(stack.pop())
        }
        return handle.invoke(ProgramStack(args))
    }

    fun call(arg: Any): Any {
        if (arity != 1) {
            throw RuntimeException("Invalid arity")
        }
        return handle.invoke(ProgramStack(arg))
    }

    fun call(vararg args: Any): Any {
        if (arity != args.size) {
            throw RuntimeException("Invalid arity")
        }
        return handle.invoke(ProgramStack(args))
    }
}